### PR TITLE
Add more FREB events/ tests for FREB events

### PR DESCRIPTION
**Pull request by jkotalik**
Recommend no white space: https://github.com/aspnet/IISIntegration/pull/1461?w=1

- Adds ANCM events to both mof file and aspnetcore_events.h (note aspnetcore_events.h is auto generated from the mof file)
- Adds tests for basic events (I think I can add client disconnect via a single process test, will check tomorrow)
- Cleanup calls to ANCMEvents in inprocesshandler.

### CODE REVIEW

**Comment by jkotalik** [comment]
@natemcmaster seems to be some sign tool issues with the M.A.S.IISIntegration package.

2018-10-02T01:26:09.2744709Z signtool : error : D:\a\1\s\artifacts\build\Microsoft.AspNetCore.Server.IISIntegration.2.2.0-preview3-20181001.2.nupkg has part lib/netstandard2.0/Microsoft.AspNetCore.Server.IISIntegration.dll which does not match the content in the binaries directory [D:\a\.dotnet\buildtools\korebuild\2.2.0-preview1-20180928.5\KoreBuild.proj]
2018-10-02T01:26:09.2744958Z signtool : error : D:\a\1\s\artifacts\build\Microsoft.AspNetCore.Server.IISIntegration.2.2.0-preview3-20181001.2.symbols.nupkg has part lib/netstandard2.0/Microsoft.AspNetCore.Server.IISIntegration.dll which does not match the content in the binaries directory [D:\a\.dotnet\buildtools\korebuild\2.2.0-preview1-20180928.5\KoreBuild.proj]
2

**Comment by natemcmaster** [comment]
Yeah, the aspnet/Razor repo had similar issues. Try reverting 2fea246b7142d2883c376345816572bb432b50e3.  The bot did something bad, and we need to investigate more.

File: src/AspNetCoreModuleV2/AspNetCore/proxymodule.cpp [hunk 1]
```diff
@@ -169,5 +169,8 @@ ASPNET_CORE_PROXY_MODULE::OnAsyncCompletion(
 
 void ASPNET_CORE_PROXY_MODULE::NotifyDisconnect() const
 {
-    m_pHandler->NotifyDisconnect();
+    if (m_pHandler != nullptr)
```
Comment (mikeharder) [inline src/AspNetCoreModuleV2/AspNetCore/proxymodule.cpp]: For clarity, is this fix related to FREB or just another bug you noticed?

File: src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cpp [hunk 2]
```diff
@@ -119,10 +120,18 @@ CreateApplication(
         if (!FAILED_LOG(hr = IN_PROCESS_APPLICATION::Start(*pServer, *pHttpApplication, pParameters, nParameters, inProcessApplication)))
         {
             *ppApplication = inProcessApplication.release();
+            if (ANCMEvents::ANCM_START_APPLICATION_SUCCESS::IsEnabled(pServer->GetTraceContext()))
```
Comment (mikeharder) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cpp]: Would it make sense to have a single helper method like `RaiseEventIfEnabled()`?

File: test/Common.FunctionalTests/Inprocess/FrebTests.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        private readonly PublishedSitesFixture _fixture;
+
+        public FrebTests(PublishedSitesFixture fixture)
+        {
+            _fixture = fixture;
+        }
+
+        [ConditionalTheory]
+        [RequiresIIS(IISCapability.FailedRequestTracingModule)]
+        [InlineData("/HelloWorld", "Verbose", "ANCM_EXECUTE_REQUEST_START", "ANCM_EXECUTE_REQUEST_ASYNC", "ANCM_EXECUTE_REQUEST_END")]
```
Comment (mikeharder) [inline test/Common.FunctionalTests/Inprocess/FrebTests.cs]: Add tests for `SHUTDOWN` and `DISCONNECT`?

File: src/AspNetCoreModuleV2/InProcessRequestHandler/InProcessRequestHandler.vcxproj [hunk 4]
```diff
@@ -224,6 +224,7 @@
     </Link>
   </ItemDefinitionGroup>
   <ItemGroup>
+    <ClInclude Include="..\CommonLib\ModuleTracer.h" />
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/InProcessRequestHandler.vcxproj]: hm

File: src/AspNetCoreModuleV2/CommonLib/ModuleTracer.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+#include "aspnetcore_event.h"
+
+class ModuleTracer
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/CommonLib/ModuleTracer.h]: Take IHttpTraceContext in ctor and make all methods instance?

File: test/Common.FunctionalTests/AppHostConfig/IIS.config [hunk 6]
```diff
@@ -707,21 +707,39 @@
...
-
-            <traceProviderDefinitions>
-                <add name="WWW Server" guid="{3a2a4e84-4c21-4981-ae10-3fda0d9b0f83}">
-                    <areas>
-                        <add name="ANCM" value="65536" />
-                    </areas>
-                </add>
-            </traceProviderDefinitions>
-
+          <traceFailedRequests />
```
Comment (pakrym) [inline test/Common.FunctionalTests/AppHostConfig/IIS.config]: Do you need this with all the code in IISDeploymentParameterExtensions.cs?

File: src/AspNetCoreModuleV2/AspNetCore/ancm.mof [hunk 7]
```diff
@@ -186,12 +186,28 @@ class ANCMWinHttpCallBack:ANCM_Events
...
 
 [Dynamic,
- Description("Inprocess executing request failure") : amended,
- EventType(7),
- EventLevel(2),
- EventTypeName("ANCM_EXECUTE_REQUEST_FAIL") : amended
+ Description("Starting inprocess execute request") : amended,
+ EventType(8),
+ EventLevel(4),
+ EventTypeName("ANCM_EXECUTE_REQUEST_START") : amended
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/AspNetCore/ancm.mof]: ANCM_INPROC_

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/XElementExtensions.cs [hunk 8]
```diff
@@ -43,5 +43,15 @@ public static XElement GetOrAdd(this XElement element, string name, string attri
 
             return existing;
         }
+
+        public static XElement GetOrAddChildElementByAttribute(this XElement element, string elementName, string attributeName, string attributeValue)
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/XElementExtensions.cs]: Looks like GetOrAdd overload we already have.

File: test/Common.FunctionalTests/Inprocess/FrebTests.cs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        private void AssertFrebLogs(IISDeploymentResult result, params string[] expectedEvents)
+        {
+            var folderPath = Helpers.GetFrebFolder(result);
+
+            // Freb logs can take up to a few seconds to be created, so we retry with quite a forgiving
+            // delay.
+            var passed = false;
+            RetryHelper.RetryOperation(() =>
```
Comment (pakrym) [inline test/Common.FunctionalTests/Inprocess/FrebTests.cs]: Does this retry helper ever throw?

File: test/Common.FunctionalTests/Utilities/Helpers.cs [hunk 10]
```diff
@@ -23,6 +23,8 @@ public static class Helpers
         private static readonly TimeSpan RetryRequestDelay = TimeSpan.FromMilliseconds(100);
         private static readonly int RetryRequestCount = 10;
 
+        public static object GetFrebLogOutput { get; internal set; }
```
Comment (pakrym) [inline test/Common.FunctionalTests/Utilities/Helpers.cs]: ???

File: test/Common.FunctionalTests/Inprocess/FrebTests.cs (hunk 9 above)
Comment (jkotalik) [reply test/Common.FunctionalTests/Inprocess/FrebTests.cs]: no, hence I set a bool after success.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 11]
```diff
@@ -34,40 +34,20 @@ REQUEST_NOTIFICATION_STATUS
...
     }
     else if (m_pApplication->QueryBlockCallbacksIntoManaged())
     {
         return ServerShutdownMessage();
     }
 
-    return m_pRequestHandler(this, m_pRequestHandlerContext);
+    auto status = m_pRequestHandler(this, m_pRequestHandlerContext);
+
+    return ModuleTracer::ExecuteRequestEnd(m_pW3Context->GetTraceContext(), status);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Don't do this. Just log and return status as separate statements.

File: test/Common.FunctionalTests/Inprocess/FrebTests.cs (hunk 9 above)
Comment (pakrym) [reply test/Common.FunctionalTests/Inprocess/FrebTests.cs]: Ugh.

Comment (jkotalik) [reply test/Common.FunctionalTests/Inprocess/FrebTests.cs]: I can make a new helper that throws on the last call.

Comment (pakrym) [reply test/Common.FunctionalTests/Inprocess/FrebTests.cs]: https://github.com/aspnet/IISIntegration/blob/release/2.2/test/IISExpress.FunctionalTests/OutOfProcess/MultipleAppTests.cs#L81

File: test/Common.FunctionalTests/Inprocess/FrebTests.cs [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        public async Task CheckFrebDisconnect()
+        {
+            var parameters = _fixture.GetBaseDeploymentParameters(publish: true);
+            parameters.EnableFreb("Verbose");
+
+            var result = await DeployAsync(parameters);
+
+            using (var connection = new TestConnection(result.HttpClient.BaseAddress.Port))
+            {
+                await connection.Send(
```
Comment (pakrym) [inline test/Common.FunctionalTests/Inprocess/FrebTests.cs]: This would be racy, the request might not get to the handler before you disconnect.

File: test/Common.FunctionalTests/AppHostConfig/IIS.config [hunk 13]
```diff
@@ -707,21 +707,16 @@
         </staticContent>
 
         <tracing>
-
-            <traceFailedRequests />
-
-            <traceProviderDefinitions>
+              <traceFailedRequests />
```
Comment (pakrym) [inline test/Common.FunctionalTests/AppHostConfig/IIS.config]: Revert?

File: test/Common.FunctionalTests/Inprocess/FrebTests.cs [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            dict.Add("Async End", new string[] { "HelloWorld", "ANCM_INPROC_ASYNC_COMPLETION_COMPLETION", "0" });
+            dict.Add("Managed End", new string[] { "HelloWorld", "ANCM_INPROC_MANAGED_REQUEST_COMPLETION" });
+            dict.Add("Request Failure", new string[] { "Throw", "ANCM_INPROC_ASYNC_COMPLETION_COMPLETION", "2" });
+            return dict;
+        }
+
+        [ConditionalTheory]
+        [RequiresIIS(IISCapability.FailedRequestTracingModule)]
+        [MemberData(nameof(FrebScenarios))]
+        public async Task CheckCommonFrebEvents(string scenario)
```
Comment (pakrym) [inline test/Common.FunctionalTests/Inprocess/FrebTests.cs]: You are deploying 5 times to test 1 scenario ("HelloWorld"). I think it's a bit wasteful.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 15]
```diff
@@ -27,47 +27,29 @@ IN_PROCESS_HANDLER::IN_PROCESS_HANDLER(
    m_pAsyncCompletionHandler(pAsyncCompletion),
    m_pDisconnectHandler(pDisconnectHandler)
 {
+    m_moduleTracer = std::make_unique<ModuleTracer>(pW3Context->GetTraceContext());
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Don't make it a reference just a value field.

File: test/Common.FunctionalTests/Inprocess/FrebTests.cs [hunk 16]
```diff
@@ -81,37 +81,37 @@ public async Task CheckFrebDisconnect()
...
+            Directory.CreateDirectory(_logFolderPath);
+            var result = await DeployAsync(parameters);
+            return result;
+        }
 
+        private void AssertFrebLogs(IISDeploymentResult result, params string[] data)
+        {
+            var folderPath = Helpers.GetFrebFolder(_logFolderPath, result);
+            // only get the last freb log.
+            var fileString = Directory.GetFiles(folderPath).Where(f => f.EndsWith("xml")).Last();
```
Comment (pakrym) [inline test/Common.FunctionalTests/Inprocess/FrebTests.cs]: Last in what order? I don't think GetFiles is ordered.

**Review by pakrym — APPROVED** [review]
